components/scream/src/share/util/scream_column_ops.hpp

```diff
@@ -79,218 +87,102 @@ class ColumnOps {
+  {
+    compute_midpoint_values<CombineMode::Replace>(team,num_mid_levels,x_i,x_m,alpha,beta);
+  }
+  // Compute X at level midpoints, given X at level interfaces
+  template<CombineMode CM, typename InputProvider, typename ScalarT>
+  KOKKOS_INLINE_FUNCTION
+  static void
+  compute_midpoint_values (const TeamMember& team,
+                           const int num_mid_levels,
+                           const InputProvider& x_i,
+                           const view_1d<ScalarT>& x_m,
                            const scalar_type alpha = one(),
                            const scalar_type beta = zero())
   {
     // Sanity checks
     debug_checks<InputProvider>(num_mid_levels,x_m);
 
-    // For GPU (or any build with pack size 1), things are simpler
-    if (pack_size==1) {
-      team_parallel_for(team,num_mid_levels,
-                        [&](const int& k) {
-        auto tmp = ( x_i(k) + x_i(k+1) ) / 2.0;
-        combine<CM>(tmp, x_m(k), alpha, beta);
-      });
-    } else {
-
-      const auto NUM_MID_PACKS     = pack_info::num_packs(num_mid_levels);
-      const auto NUM_INT_PACKS     = pack_info::num_packs(num_mid_levels+1);
-
-      // It is easier to read if we check whether #int_packs==#mid_packs.
-      // This lambda can be used in both cases to process packs that have a next pack
-      auto shift_and_avg = [&](const int k){
-        // Shift's first arg is the scalar to put in the "empty" spot at the end
-        pack_type tmp = ekat::shift_left(x_i(k+1)[0], x_i(k));
-        tmp += x_i(k);
-        tmp /= 2.0;
-        combine<CM>(tmp, x_m(k), alpha, beta);
-      };
-      if (NUM_MID_PACKS==NUM_INT_PACKS) {
-        const auto LAST_PACK =  NUM_MID_PACKS-1;
-
-        // Use SIMD operations only on NUM_MID_PACKS-1, since mid pack
-        // does not have a 'next' one
-        team_parallel_for(team,NUM_MID_PACKS-1,shift_and_avg);
-
-        team_single (team, [&]() {
-          // Last level pack treated separately, since int pack k+1 does not exist.
-          // Shift's first arg is the scalar to put in the "empty" spot at the end.
-          // In this case, we don't need it, so just uze zero.
-          const auto& xi_last = x_i(LAST_PACK);
-          pack_type tmp = ekat::shift_left(zero(), xi_last);
-
-          tmp += xi_last;
-          tmp /= 2.0;
-          combine<CM>(tmp, x_m(LAST_PACK), alpha, beta);
-        });
-      } else {
-        // We can use SIMD operations on all NUM_MID_PACKS mid packs,
-        // since x_i(k+1) is *always* fine
-        team_parallel_for(team,NUM_MID_PACKS,shift_and_avg);
-      }
-    }
+    compute_midpoint_values_impl<CM,InputProvider,ScalarT>(team,num_mid_levels,x_i,x_m,alpha,beta);
   }
 
   // Compute X at level interfaces, given X at level midpoints and top or bot bc.
   // Note: with proper bc, then x_int(x_mid(x_int))==x_int.
-  template<bool FixTop, typename InputProvider = DefaultProvider>
+  template<bool FixTop, typename InputProvider, typename ScalarT>
```

Comment:
Is it worth describing the FixTop template parameter and what it does, in a comment?